### ISSUE TITLE
an example that shows the need for memory backpressure

### ISSUE THREAD

**Issue by rabernat** [dask/distributed#2602]
In my work with large climate datasets, I often concoct calculations that cause my dask workers to run out of memory, start dumping to disk, and eventually grind my computation to a halt. There are many ways to mitigate this by e.g. using more workers, more memory, better disk-spilling settings, simpler jobs, etc. and these have all been tried over the years with some degree of success. But in this issue, I would like to address what I believe is the root of my problems within the dask scheduler algorithms.

The core problem is that the tasks early in my graph generate data faster than it can be consumed downstream, causing data to pile up, eventually overwhelming my workers. Here is a self contained example:

```python
import dask.array as dsa

# create some random data
# assume chunk structure is not under my control, because it originates
# from the way the data is laid out in the underlying files
shape = (500000, 100, 500)
chunks = (100, 100, 500)
data = dsa.random.random(shape, chunks=chunks)

# now rechunk the data to permit me to do some computations along different axes
# this aggregates chunks along axis 0 and dis-aggregates along axis 1
data_rc = data.rechunk((1000, 1, 500))
FACTOR = 15


def my_custom_function(f):
    # a pretend custom function that would do a bunch of stuff along
    # axis 0 and 2 and then reduce the data heavily
    return f.ravel()[::15][None, :]

# apply that function to each chunk
c1 = math.ceil(data_rc.ravel()[::FACTOR].size / c0)
res = data_rc.map_blocks(my_custom_function, dtype=data.dtype,
                         drop_axis=[1, 2], new_axis=[1], chunks=(1, c1))

res.compute()
```

(Perhaps this could be simplified further, but I have done my best to preserve the basic structure of my real problem.)

When I watch this execute on my dashboard, I see the workers just keep generating data until they reach their memory thresholds, at which point they start writing data to disk, before `my_custom_function` ever gets called to relieve the memory buildup. Depending on the size of the problem and the speed of the disks where they are spilling, sometimes we can recover and manage to finish after a very long time. Usually the workers just stop working.

This fail case is frustrating, because often I can achieve a reasonable result by just doing the naive thing:
```
for n in range(500):
    res[n].compute()
```
and evaluating my computation in serial.

I wish the dask scheduler knew to stop generating new data before the downstream data could be consumed. I am not an expert, but I believe the term for this is [backpressure](https://www.quora.com/What-is-backpressure-in-the-context-of-data-streaming). I see this term has come up in https://github.com/dask/distributed/issues/641, and also in [this blog post](http://matthewrocklin.com/blog/work/2017/04/13/streaming) by @mrocklin regarding streaming data.

I have a hunch that resolving this problem would resolve many of the pervasive but hard-to-diagnose problems we have in the xarray / pangeo sphere. But I also suspect it is not easy and requires major changes to core algorithms.

Dask version 1.1.4

**Comment by TomAugspurger** [comment]
I’ll be on paternity leave for at least the next two weeks (maybe longer, depending on how things are going). If this is still open then I’ll tackle it then.

**Comment by rabernat** [comment]
Hi folks...I'm just pinging this issue to remind us that it remains a crucial priority for Pangeo.

**Comment by TomAugspurger** [comment]
👍 it's on my list for tomorrow :)

**Comment by TomAugspurger** [comment]
https://nbviewer.jupyter.org/gist/TomAugspurger/2df7828c22882d336ad5a0722fbec842 has a few initial thoughts. The first problem is that the rechunking from along just axis 0 to just axis 1 cause a tough, global communication pattern. The output of `my_custom_function` will need input from every original chunk.

![image](https://user-images.githubusercontent.com/1312546/58256320-6b17bd00-7d34-11e9-959a-dade6c5cebaf.png)

So *for this specific problem* it *may* be an option to preserve the chunks along axis 0, and only add additional chunks along axis 1. This leads to a better communication pattern.

![image](https://user-images.githubusercontent.com/1312546/58256444-a2866980-7d34-11e9-9b39-9de8e72dd7cc.png)

But, a few issues with that

1. It may not work for the real problem (@rabernat do you have a "real world" example available publicly somewhere?). We should still attempt a proper solution to the problem.
2. It increases the number of tasks, which may bump up against other scheduler limitations.

I'll keep looking into this, assuming that a different chunking pattern isn't feasible.

(FYI @mrocklin the HTML array repr came in handy).

**Comment by martindurant** [comment]
@TomAugspurger , the general problem remains interesting, though, of using the expected output size and known set of things that could be dropped of a given tasks as an additional heuristic in determining when/where to schedule that task.

**Comment by TomAugspurger** [comment]
Yep.

I notice now that my notebook oversimplified things. The original example still add chunks along the first and second axis (which means the communication shouldn't be entirely global). I'll update things.

**Comment by rabernat** [comment]
This is a "real world" example that can be run from ocean.pangeo.io which I believe illustrates the core issue. It is challenging because the chunks are big, leading to lots of memory pressure.

```python
import intake
cat = intake.Catalog('https://raw.githubusercontent.com/pangeo-data/pangeo-datastore/master/intake-catalogs/ocean.yaml')
ds = cat.GODAS.to_dask()
print(ds)
salt_clim = ds.salt.groupby('time.month').mean(dim='time')
print(salt_clim)

# launch dask cluster
from dask.distributed import Client
from dask_kubernetes import KubeCluster
# using more workers might alleviate the problem, but that is not a satisfactory workaround
cluster = KubeCluster(n_workers=5)
client = Client(cluster)

# computation A
# compute just one month of the climatology
# it works fine, indicating that the computation could be done in serial
salt_clim[0].load()

# computation B
# now load the whole thing
# workers quickly run out of memory, spill to disk, and even get killed
salt_clim.load()
```

![image](https://user-images.githubusercontent.com/1197350/58264566-14b67880-7d4c-11e9-8f15-dfcf29be419e.png)

**Comment by TomAugspurger** [comment]
Thanks @rabernat. I'll try it out on the pangeo binder once I have something working locally.

I think (hope) I have a workable solution. My basic plan is

1. Track the cumulative output `.nbytes` of tasks at the prefix-level (easy)
2. When deciding how to schedule under load, look up / compute the relative size
   of the input to the output again, at prefix level. Prioritize tasks that look
   like they'll reduce memory usage. (seems harder. I don't know if we use
   prefix-level relationships anywhere else in scheduling decisions).

So if we currently have high memory usage, and we see a set of tasks like


```
      a1   a2  # 5 bytes, 5 bytes   |   x1  x2   # 5 bytes, 5 bytes
        \ /                         |    \ /
         b     # 1 byte             |     y      # 20 bytes
```

We would (ideally) schedule `b` before `y`, since it tends to have a net
decrease in memory usage. Or if we have tasks running that we know will free up
some memory, we might wait to schedule `y` since it tends to increase memory.
I'm sure there are things I'm missing, but this should provide a good start.

**Comment by martindurant** [comment]
@TomAugspurger , that's exactly how I was picturing things. Additionally, you may need to know whether calculating `b`, for example, actually releases `a1` and `a2`.

**Comment by TomAugspurger** [comment]
> you may need to know whether calculating b, for example, actually releases a1 and a2.

Yeah, I've been looking through the recommendations -> release logic now. I'm wondering if we can also keep a counter of "completing this task *recommended* that we release this many bytes". That seems a bit easier to track than knowing whether `a1` and `a2` were actually released.

**Comment by rabernat** [comment]
As a user of dask.array, it confuses me that the graph itself doesn't have this information in it already. In my mind, all of my dask.array operations know about the shape and type of their inputs and outputs. Consequently, the memory impact of a task is known a-priori.

I do understand that this information is not encoded into the dask graph--all the graph knows about is tasks and their dependencies. But, to my naive interpretation, it seems harder to try to "learn" the memory impact of tasks based on past experience than it does to pass this information through the graph itself, in the form of task annotation.

**Comment by TomAugspurger** [comment]
> As a user of dask.array, it confuses me that the graph itself doesn't have this information in it already. In my mind, all of my dask.array operations know about the shape and dtype of their inputs and outputs. Consequently, the memory impact of a task is known a-priori.

As you say, TaskAnnotations mentioned by @sjperkins in https://github.com/dask/distributed/issues/2602#issuecomment-474232764 would let us pass that information through. I think that approach should be considered, but I have two reasons for pursing the "learning" approach within distributed for now

1. I suspect it's easier to get a rough implementation of this in place. Task annotations will (I think) be a larger change to dask.
2. There always be cases where the user doesn't or can't provide the expected memory usage when the graph is constructed. Say as a result of a boolean filter, or when using `delayed` on a source that can't easily provide the memory usage information.

I will take another look at the task annotation issue though.

**Comment by TomAugspurger** [comment]
A promising note, the additional LOC for the tracking seems to be ~3 :) I'm just keeping a `task_net_nbytes` that tracks the difference between task input and output at the prefix-level. With that, we correctly see that

1. random sample is a net increaser of memory
2. rechunking has no effect on memory (again, we aren't measuring transient memory here, just final).
3. my_custom_function is a net reducer of memory

```
--------------------------------------------------------------------------------
# task_net_nbytes after all the computation is done.
{'my_custom_function': -18666656,
 'random_sample': 20000000,
 'rechunk-merge': 0,
 'rechunk-split-rechunk-merge': 0}
```

The scheduling changes will be harder, but this seems promising so far.

**Comment by mrocklin** [comment]
@TomAugspurger said

> We would (ideally) schedule b before y, since it tends to have a net
decrease in memory usage

Sounds reasonable.  As a warning, there might be other considerations fighting for attention here.  Scheduling is an interesting balancing game.

@martindurant said

> Additionally, you may need to know whether calculating b, for example, actually releases a1 and a2.

This is a nice point.  One naive approach would be to divide the effect by the number of dependents of the dependency-to-be-released.

**Comment by mrocklin** [comment]
Also, I'm very glad to have more people thinking about the internal scheduling logic!

**Comment by TomAugspurger** [comment]
@mrocklin do you know off hand whether the changes to the scheduling logic are likely to be on the worker, scheduler, or both? I can imagine both situations occurring, so I suspect the answer is "both".

**Comment by mrocklin** [comment]
I would say either.  Dask has task priorities in three places:

   1. As provided by the client with dask.order and user priorities
   2. As modified by the scheduler, with first-in-first-out for successive
   compute calls
   3. As modified again by the worker, loosely preferring first-in-first
   out, but largely deferring to the high priorities

What you're proposing seems like it plays at the first level, similar to
dask.order.  You're going to want to be careful not to overwhelm the user
provided priorities.  I think that you're going to hit some interesting
situations where bytes handling and dask.order disagree.  In some ways
bytes handling is smarter on a short term basis, but might bite you by
being short sighted.  It will be interesting to see what happens.

In terms of scheduler vs worker, the scheduler would be nicer because you
can learn the bytes size across the cluster as a whole, rather than having
to relearn on every worker.  Learning on the worker is a bit nicer because
you'll be able to update priorities more rapidly and it's always nice to
move any potentially costly logic away from the scheduler.  My sense is
that it might not matter much where you implement the logic to start.

**Comment by mrocklin** [comment]
I could also imagine us not asking the client to run dask.order at all, and doing all task prioritization on the workers.  This would allow us to integrate the byte sizes into the dask.order computation (and remove it from the single-threaded client-scheduler processing).  This is probably something to explore as future work though.

**Comment by TomAugspurger** [comment]
> I could also imagine us not asking the client to run dask.order at all, and doing all task prioritization on the workers.

FWIW, `dask.order` does correctly order this workload. The final task of each parallel branch comes before the first task of the next branch.

![Unknown](https://user-images.githubusercontent.com/1312546/58502338-f2599c00-814b-11e9-989a-5bfd2c3785a8.png)

So this looks more like a problem with reordering / waiting to execute some tasks, if we notice that we

1. are at high memory
2. Have a set of ready tasks that are likely to reduce memory usage.

**Comment by TomAugspurger** [comment]
@mrocklin question for you. In `scheduler.py` when a task is completed, `_add_to_memory` handles some things including marking that tasks dependencies as (maybe) ready for release https://github.com/dask/distributed/blob/d202e6253ed8ddc7919d0d4f128d88954e9859b8/distributed/scheduler.py#L3472-L3479

Do you know if there's something similar for Worker? `Worker.put_key_in_memory` doesn't seem to have logic for maybe releasing dependencies of a recently completed task.

In my current approach, I'm using whether a task releases other tasks to measure the net memory usage of completing a task (output nbytes - sum of released memory).

**Comment by mrocklin** [comment]
Right, the Worker doesn't have the full graph (this would be hard to make consistent, especially if the graph is changing rapidly).  It signals to the scheduler that a task has finished and hopefully a short time later the scheduler sends a release message back.

**Comment by mrocklin** [comment]
OK, then maybe prioritization isn't the right place to put this.  Maybe workers just refuse to run tasks that are likely to increase memory when they are under memory pressure.

**Comment by mrocklin** [comment]
>  I'll look into a "reorder priorities while under load" method on the scheduler.

To be clear, I think that the best place to handle this is on the worker.  It will have many tasks that it can run, at some point when an individual worker runs low on free memory it switches to preferring memory-freeing tasks.

**Comment by TomAugspurger** [comment]
@rabernat in your computation, do you recall seeing messages like `WARNING Memory use on   tcp://192.168.7.20:53133 is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?` before your workers were killed?

I think the backpressure / reschedule based on memory usage idea is still valuable, but that warning, and the root cause behind it, are making things a bit more difficult for me. That warning is saying that we don't have an accurate picture of the worker processes memory, so scheduling decisions based on that are likely to be faulty. I'm trying to decide which rabbit hole to go down first.

**Comment by rabernat** [comment]
When I try this on ocean.pangeo.io with 5 workers, I don't see those errors. And I don't actually have dying workers.

What I see is a task stream that looks like this:
![image](https://user-images.githubusercontent.com/1197350/58890524-c8a6f480-86b8-11e9-8f2f-daefe74afac6.png)

At this point, all the workers are "orange" in memory consumption and are spilling massive amounts of data to disk. Progress grinds to a halt, but no errors are raised. (I interrupted the computation at this point; maybe if I had let it run longer something different would happen.)

**Comment by TomAugspurger** [comment]
Thanks, I'll try out the original example on ocean.pangeo.io (when I tried the real-world example in https://github.com/dask/distributed/issues/2602#issuecomment-495262673 last weeks things seemed to work fine with a few workers)

**Comment by rabernat** [comment]
> when I tried the real-world example in [#2602 (comment)](https://github.com/dask/distributed/issues/2602#issuecomment-495262673) last weeks things seemed to work fine with a few workers

This dataset was updated (made smaller) and probably doesn't produce that problem anymore.

**Comment by rabernat** [comment]
Here is another classic example that produces the same problem. It uses xarray, but only to produce the indices needed for grouping. This is **the example** that is most relevant for climate stuff, because it is related to calculating climatologies.

```python
import dask.array as dsa
import numpy as np
import xarray as xr

data = dsa.random.random((10000, 1000000), chunks=(1, 1000000))
da = xr.DataArray(data, dims=['time', 'x'],
                  coords={'day': ('time', np.arange(10000) % 100)})
clim = da.groupby('day').mean(dim='time')
anom = da.groupby('day') - clim
anom_mean = anom.mean(dim='time')

# works fine, meaning I could evaluate the mean in serial if need by
# by using a loop
anom[0].load()

# runs out of memory quickly
anom_mean.load()
```

If you're going to focus on one example, this is the most important one.

**Comment by mrocklin** [comment]
This is, I think, a different situation from a scheduling perspective.  I don't think that there is a way to compute that example in small memory without loading data twice.  You have to load all of the data in order to compute `clim`. And then you need that data again to compute `anom_mean`.  

The scheduling ask here is "be willing to delete and recompute data" while the scheduling ask before was "prefer tasks that release memory, and be willing to hold off a bit on tasks that increase net data".

I don't mean to say that this isn't also a valid issue, but it's completely different work on the Dask side, even though both issues manifest as memory issues.

**Comment by mrocklin** [comment]
(I also don't see a clear way forward on the "be willing to delete and recompute data" approach.  A change like that would touch a lot of the scheduling code.  I would estimate that at developer-months of time.)

**Comment by rabernat** [comment]
Matt, thanks for the clarification. From my perspective, the issues seemed very similar--both fail due to memory overconsumption but can be executed successfully in a loop. But I understand now how they are not.

**Comment by TomAugspurger** [comment]
Some potentially interesting observations below on the original example (note that I've edited the call to `map_blocks` to fix the output chunks, so that we get the metadata right. I'm also using a larger reduction (every 100th elements, rather than 15th) to make the reduced output fit on my laptop).

1. We can compute `data_rc.blocks[0, 0]` just fine (~4 Mb). The custom reduction reduces the 4 Mb -> 40 Kb. We have ~500 of these per *output* block,
   So `[500 * 4 Kb] -> 4 Mb` per block in `res`.
2. We *cannot* compute `res.blocks[0, 0]`, let alone `res.compute()`.
3. We *can* do things manually, roughly

```python
# manually build the graph

vblocks = []
for i in range(data_rc.numblocks[0]):
    b = data_rc.blocks[0]
    c1 = b.ravel()[::FACTOR].size // b.numblocks[1]
    r = b.map_blocks(lambda x: x.ravel()[::FACTOR][None, :], drop_axis=2, chunks=(1, c1))
    vblocks.append(r)

r = da.concatenate(vblocks)
```

Together, 2 and 3 point to an issue with `map_blocks` not generating the ideal task graph when dropping axes and compressing chunks. I'll look into that independently. I think this scheduling issue is still higher priority.

When replacing the `data_rc.map_blocks(my_custom_function, ...)` with the manually built version, I'm able to successfully perform the 200 Gb -> 2 Gb reduction on my laptop (2 workers, 2.5Gb each).

**Comment by rabernat** [comment]
> I also don't see a clear way forward on the "be willing to delete and recompute data" approach.

Would it be possible for the user to "trick" dask into recomputing the data by using distinct tokens?

**Comment by TomAugspurger** [comment]
I think the current way for users to work around this is to persist the small output of the large computation first:

```python
clim = da.groupby('day').mean(dim='time').persist()  # or .compute()

anom = da.groupby('day') - clim
anom_mean = anom.mean(dim='time')
```

That's essentially manually doing the two passes over the data that's required for this computation to not blow memory.

**Comment by sjperkins** [comment]
How about registering a memory estimation function for a task? The idea being that, given the inputs to the task, it would estimate the task's memory output.

The idea is somewhat similar to tensorflow's shape estimation API - given an operator's input shapes one can try to infer the output shape.

**Comment by TomAugspurger** [comment]
> How about registering a memory estimation function for a task?

FWIW, with #2765 something like this would be possible with a scheduler plugin (but not documented / officially supported). I'm hesitant to do something like this now, since I think that Task annotations are likely to occur someday.

**Comment by sjperkins** [comment]
> FWIW, with #2765 something like this would be possible with a scheduler plugin (but not documented / officially supported). I'm hesitant to do something like this now, since I think that Task annotations are likely to occur someday.

OK, just to be clear I wasn't necessarily thinking of including it under the banner of the Task Annotations. More something along the lines of the following pseudo-code:

```python
def my_sum(a, b):
   """ Sum equally shaped a and b arrays """
   return a + b

@register_mem_estimator(my_sum)
def _sum_estimator(a, b):
    """ Reduction, net sink """
    return a.nbytes

class Scheduler(object):
   ...
   def schedule(fn, *args, **kwargs):
       try:
           mem_estimator = estimator_lookup[fn]
       except ImportError:
           out_bytes = heuristic_estimator(fn, *args, **kwargs)
       else:
           out_bytes = mem_estimator(*args, **kwargs)
```

However, I don't understand the scheduler well enough to understand if the above is possible or any of the possible constraints.

**Comment by mrocklin** [comment]
> Would it be possible for the user to "trick" dask into recomputing the data by using distinct tokens?

You could try fusing these functions into other tasks.  We do this today with some getitem functinos

https://github.com/dask/dask/blob/6e8c1b76feb12337574f40032dbd3818626b8e28/dask/array/optimization.py#L26

**Comment by MikeAlopis** [comment]
Hello @TomAugspurger @mrocklin any updates on progress here? Thank you !

**Comment by TomAugspurger** [comment]
Sorry, I was off last week. Getting back to it this week. But a short status update.

This branch does two things

1. Gradually learns the likely net memory usage of completing a task, by observing the memory use of the output less the memory use of all freed dependencies.
2. Adjust the Worker scheduling logic to prioritize memory-freeing tasks when we notice that we're under memory pressure.

Part 1 works well enough for now, I think (task annotations may help make things more exact in the future).

I'm struggling with part 2. I'm having trouble finding an acceptable scheduling policy that doesn't totally kill performance in some workloads.

@MikeAlopis if you have additional examples where you're unexpectedly running into worker memory errors, I'd be happy to see them. I think I'll want to test this out on a diverse set of workloads.

**Comment by mrocklin** [comment]
Would it make sense to merge in part 1 before part 2?

**Comment by TomAugspurger** [comment]
> Would it make sense to merge in part 1 before part 2?

For review ease of review, yes. https://github.com/dask/distributed/pull/2847

**Comment by TomAugspurger** [comment]
I've been poking at this again. In the examples I've been using, the root cause of the memory blowing up is communication between workers. I see a few ways to address this

1. Avoid the communication in the first place. This can (sometimes) be done by adjusting the scheduling. Dask's depth-first scheduling isn't the best for work patterns like
```
        a-1   a-2    a-3  a-4
          \   /       \   /
           b-1         b-2
            |           |
           c-1         c-2
```
With two machines, we might schedule `a-1` on machine 1 and `a-2` on machine 2. Then we'll need to move the result of one to the other machine. We (maybe) would want to schedule `a-1` and `a-2` on the same machine. I have some thoughts on an API here, but haven't coded anything up yet. This wouldn't be a general solution, since many problems require some communication.
2. Reduce (or disable) overlapping communication & compute, at least when memory-use is high. This is essentially the backpressure request, limited to "don't schedule new tasks when we're communicating".

**Comment by mrocklin** [comment]
> We (maybe) would want to schedule a-1 and a-2 on the same machine

Indeed.  This would be great, and would help to accelerate many workloads if we could make this general.  

> I have some thoughts on an API here, but haven't coded anything up yet. 

I'm mildly against a short term solution here that accepts hints from a user.  Historically we've solved all of these scheduling problems at the core so that they affect all uses, not a special few.  This approach has been, so far, effective.

If we can find a way to make it cheap enough, something might find its way into the worker objective: 

https://github.com/dask/distributed/blob/e02cc4409e352e40e4128fc542b8aaed51b5a01f/distributed/scheduler.py#L4560-L4575

There are things to balance here.  It's an interesting problem.

**Comment by TomAugspurger** [comment]
`worker_objective` *might* be too late, at least for the kinds of problems I'm playing with right now.

For problems that start with many inputs, which can be processed (mostly) in parallel, we go down the not-ideal path in the `elif self.idle` here

https://github.com/dask/distributed/blob/e02cc4409e352e40e4128fc542b8aaed51b5a01f/distributed/scheduler.py#L3612-L3630

Naively that makes sense. But for certain problems we're willing to pay a bit of time to find a better worker for each tasks.

@mrocklin do you know if we have any scheduling logic that looks at the "siblings" of a task? If I have (top to bottom)

```
        a-1   a-2  a-3   a-4  a-5  a-6
          \    |    /     \    |    /
              b-1             b-2
```

When I'm scheduling `a-2`, I'd like to look at where `a-1` and `a-3` are perhaps scheduled to run (in testing on a toy problem, assigning this task to the worker with the most sibling tasks works).

**Comment by mrocklin** [comment]
It gets even stranger, right?  Not just siblings, but cousins of various levels of removed-ness :)  And at some point you're going to want to cut the family tree.

I don't know of any scheduling logic that we have currently.  If I were to start solving this problem I would start with pen and paper, write down a variety of situations, figure out how I would go about scheduling them in a few situations, and hopefully build out some intuition for the problem.

**Comment by dougiesquire** [comment]
Any updates or progress here? I’ve been experiencing symptoms that may be related to the same issue.

When running a `groupby.mean` operation on a relatively large dataset using a dask cluster the memory usage of each worker increases very rapidly, to values substantially larger than what I would expect based on the chunking. The issue is particularly severe when accessing and processing data stored on an ultra-fast parallel file storage system.

My example also uses xarray,
![Screen Shot 2019-09-25 at 10 04 25 am](https://user-images.githubusercontent.com/42455466/65558919-e2c45580-df7b-11e9-9f2e-a1001cddcf2d.png)

Interestingly, the amount of memory used per worker is a function of the number of workers I provide. For a single worker, the maximum memory load is about what I would expect for this persist (~45GB). For two workers, each max out at 80GB per worker before the calculation finishes. For three workers, they start to spill to disk before the calculation completes (120GB per worker is the maximum I can request on the system I am using).

**Comment by rabernat** [comment]
Just chiming in to say this remains a persistent and serious problem for many pangeo users. We are keen to help resolve it but not sure the best way to engage.

**Comment by mrocklin** [comment]
> We are keen to help resolve it but not sure the best way to engage.

I get the sense that at this point the right way to engage is to find people who are interested in low memory task scheduling and have them dive into Dask's heuristics.

@TomAugspurger you looked at this problem most recently.  Are you able to summarize the current state of things and avenues for work as you see them?

**Comment by TomAugspurger** [comment]
I'm still poking at it, but all my attempts (https://github.com/dask/distributed/pull/2847, https://github.com/dask/distributed/pull/2940, and others that haven't made it to PR stage) have failed.

Thus far, it's seemed like the most common issues have been communication heavy workloads. So the two solutions I've attempted are better scheduling to not need as much communication, and better handling communication when we are already high on memory.

Collecting workloads that fail is still helpful.

**Comment by TomAugspurger** [comment]
Turning to @dougiesquire's example (thanks for posting that), I'm roughly trying to replicate it with the following random data:

```python
size = (3, 24, 96, 5, 9, 10)
chunks = (1, 1, 96, 5, 9, 10)
arr = da.random.random(size, chunks=chunks)
arr

items = dict(
    ensemble = np.arange(size[0]),
    init_date = pd.date_range(start='1960', periods=size[1]),
    lat = np.arange(size[2]).astype(float),
    lead_time = np.arange(size[3]),
    level = np.arange(size[4]).astype(float),
    lon = np.arange(size[5]).astype(float),
)
dims, coords = zip(*list(items.items()))

array = xr.DataArray(arr, coords=coords, dims=dims)
dset = xr.Dataset({'data': array})
dset
<xarray.Dataset>
Dimensions:    (ensemble: 3, init_date: 24, lat: 96, lead_time: 5, level: 9, lon: 10)
Coordinates:
  * ensemble   (ensemble) int64 0 1 2
  * init_date  (init_date) datetime64[ns] 1960-01-01 1960-01-02 ... 1960-01-24
  * lat        (lat) float64 0.0 1.0 2.0 3.0 4.0 ... 91.0 92.0 93.0 94.0 95.0
  * lead_time  (lead_time) int64 0 1 2 3 4
  * level      (level) float64 0.0 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0
  * lon        (lon) float64 0.0 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0
Data variables:
    data       (ensemble, init_date, lat, lead_time, level, lon) float64 dask.array<chunksize=(1, 1, 96, 5, 9, 10), meta=np.ndarray>
```

**Comment by dougiesquire** [comment]
Stupid question:
My solution to issues of this type at the moment is to use a single, multi-core worker. This works great for fixing the memory issue, but obviously means I'm limited by the size of a single node. Is there a way to "turn off" inter-worker communication?

**Comment by TomAugspurger** [comment]
For now, your best bet may be set the `distributed.worker.connections.outgoing` config value to some small number (3-10?)

https://github.com/dask/distributed/pull/3071 will effectively do that automatically when the worker is under high memory pressure, so that too much communication won't end up killing a high-memory worker, but you still get the benefit of communicating with many workers when memory is normal.

**Comment by dougiesquire** [comment]
> For now, your best bet may be set the `distributed.worker.connections.outgoing` config value to some small number (3-10?)

Thanks for this, @TomAugspurger. Unfortunately tweaking `distributed.worker.connections.outgoing` doesn't seem to make much difference at all.

**Comment by TomAugspurger** [comment]
@dougiesquire what does the dashboard show just before the workers are dying? High memory usage and a lot of communication (overlaid in red)?

**Comment by dougiesquire** [comment]
> @dougiesquire what does the dashboard show just before the workers are dying? High memory usage and a lot of communication (overlaid in red)?

Yup - this is a screen shot of the dashboard as the workers start to die
![Screen Shot 2019-10-02 at 8 10 25 am](https://user-images.githubusercontent.com/42455466/66005060-b9f90e80-e4ed-11e9-92d2-ad25b5711779.png)

**Comment by rabernat** [comment]
> the presence of a single root task that's the dependency of all the `open_zarr` tasks causes issues with optimization

What is this "root task"? Is it something we could eliminate on the xarray (or dask) side?

**Comment by mrocklin** [comment]
> What is this "root task"?

By this I mean a task on which many other things depend.  It is probably a reference to the Zarr file.  You could probably optionally recreate the Zarr file in every task.  In some cases that might slow things down.  I'm not sure.

**Comment by mrocklin** [comment]
Folks here might want to take a look at https://github.com/dask/dask/pull/5451

Tom and I were playing with high level fusion, and this was an interesting outcome.

**Comment by Thomas-Moore-Creative** [comment]
I believe @dougiesquire has provided some example(s) but is there anything additional someone facing this issue could log or document that could assist with a solution?

**Comment by mrocklin** [comment]
Nothing comes to mind at the moment.  This probably needs someone to think hard about the problem and general task scheduling policies for a while.  Most of the people who do that regularly are pretty backed up these days.

**Comment by TomAugspurger** [comment]
Short summary a call with @dougiesquire and @Thomas-Moore-Creative yesterday, looking at the problem described in https://github.com/dask/distributed/issues/2602#issuecomment-534795074.

This is a common

1. read from zarr
2. group by month
3. reduce

workload.

We eventually got things working well by setting `fuse_ave_width` to a sufficiently large value for the workload. We tried with Dask 2.9 and unfortunately the root fusion wasn't enough to fuse things correctly for this workload. I'm going to explore that a bit more today.

**Comment by Thomas-Moore-Creative** [comment]
@TomAugspurger - thank you for taking the time to look at the problem with us.  Greatly appreciated and part of why we love this Pangeo community.

**Comment by eriknw** [comment]
I can narrowly comment on @TomAugspurger's example of "`open_zarr` with `fuse_ave_width=200`" that resulted in this DAG:
![zarr-fused](https://user-images.githubusercontent.com/1312546/65603227-e4395200-df6a-11e9-9b39-716bb69595c1.png)

Specifically, this doesn't fuse down similar to "random data with fusion" like
![random-fused](https://user-images.githubusercontent.com/1312546/65603290-029f4d80-df6b-11e9-8e81-a94ea289758f.png)
because of the `hold_keys` optimization introduced in https://github.com/dask/dask/pull/2080.

Without `hold_keys`, the optimized DAG looks like
![zarr-fused](https://user-images.githubusercontent.com/2058401/71291452-50441000-2338-11ea-98a5-d628ffcd2906.png)

To be honest, I still don't fully grok what we want to accomplish with `hold_keys`, if it's still necessary, or if it can be improved.

**Comment by mrocklin** [comment]
> To be honest, I still don't fully grok what we want to accomplish with hold_keys, if it's still necessary, or if it can be improved.

Often we don't want to inline the array object into every task.  This is the case with numpy arrays (they're faster to serialize on their own, and we don't want to copy the full array for every individual task), and for objects like `h5py.Datasets` which don't work well with pickle.  It may be that with Zarr we *do* want to include the Zarr object within every task individually.  This would be the case if it was very cheap to serialize.

**Comment by eriknw** [comment]
I don't see why `hold_keys` is necessary for any of what you just said.  This may mean I don't understand what you just said, nor do I understand the previous discussions around `hold_keys`.  I understand what the code `hold_keys` is doing, but I don't get why.

Furthermore, I suspect the graph from `open_zarr` with `fuse_ave_width=200` likely represents a failure of graph optimization, and, for this, `hold_keys` is the culprit.  Whatever the goal of `hold_keys`, I suspect it's better done as part of `fuse`.  I would prefer to fix this optimization as a more general solution instead of repeating `open_zarr` in every task.

`open_zarr` above won't get fused, and it wouldn't get fused if it were a numpy array either.

It seems there are a couple considerations regarding what `hold_keys` tries to do with getters such as `getitem`.
  1. getters often return much smaller objects, so we'd sometimes rather send around the object after the getters are applied than before, and
  2. the objects getters are applied to may not be serializable, so, once again, we'd rather send the object after the getters are applied.

Is this correct?  Is (2) a primary concern?  The consequence of being unserializable is that all direct dependents must be executed on the same worker, right?

I believe the `open_zarr` example above is a failure case, because it results in sending the original objects to many, many tasks, and then the results of those tasks need to be sent around as part of a reduction.  If the user knows their workflow and sets `fuse_ave_width` very high, then dask should respect that and use the graph with three branches that I shared in my previous post.

I think `fuse` can be improved if it's given two sets of keys that:
  1. are likely to reduce the size of data, and
  2. are unable to be serialized.

Thoughts?

**Comment by mrocklin** [comment]
> Whatever the goal of hold_keys, I suspect it's better done as part of fuse

That could be.  Looking at the docstring of hold_keys, one of the main objectives was to avoid fusing things like `"x"` below:

```python
{"x": np.array([1, 2, 3, ...]),
 "y": (inc, "x"),
 "z": (dec, "x"),
  ...
}
```

We don't want `"x"` to be fused anywhere because it is easier to serialize as a raw value, rather than tucked away into a task.

> Furthermore, I suspect the graph from open_zarr with fuse_ave_width=200 likely represents a failure of graph optimization, and, for this, hold_keys is the culprit. 

I definitely agree that a width of 200 is a fail case.  What are we trying to achieve with that example?  I'm not sure that there is an optimization to be made?

For root fusion (I'm not sure that this is what we're talking about here or not) we would want to benefit in cases like the following:

```python
x = da.open_zarr(...)
y = da.open_zarr(...)
z = x + y
z.sum().compute()
```

Ideally the matching chunks from x and y would be loaded on the same machine.  This is easy to guarantee if they are part of the same task.  Root fusion does this *if* the task doesn't have any dependencies, which in this case, they do, which blocks things.  I'm not sure that any of the problems presented above have this structure though, so that might not be what we're talking about here.  

So, do we know what the ask is here?  Do we have confidence that the problem presented above is likely to be resolved by improved optimization?

> Is (2) a primary concern? The consequence of being unserializable is that all direct dependents must be executed on the same worker, right?

Not quite.  We *can* serialize these things, but only if they're bare tasks, because then our custom serialization can take over, rather than having to rely on pickle.

**Comment by takluyver** [comment]
We've been seeing a number of weird errors (like killed workers) for which I think this problem might be the root cause.

We create dask arrays from HDF5 files, stack/concatenate them together to make one big array, and then typically do some slicing and dicing before a reduction step. We're using `dask_jobqueue` to launch workers through Slurm. As the data involved gets bigger, I believe it's trying to load data too quickly rather than pushing it through to the reduction step. What puzzles me is: what are we doing that's strange? It seems like exactly the kind of use case that dask.array is meant to handle, and clearly not everyone has this problem.

**Comment by rabernat** [comment]
> We create dask arrays from HDF5 files, stack/concatenate them together to make one big array, and then typically do some slicing and dicing before a reduction step.

This is essentially the same pattern as climate science.

> It seems like exactly the kind of use case that dask.array is meant to handle, and clearly not everyone has this problem.

Many of us do indeed have this problem. Users have developed several workarounds, none very satisfactory:

- Throw more memory / larger cluster at the problem, which sometimes makes it go away
- Allow workers to spill to a fast local disk (SSD), which acts like extra memory
- Split the job up into smaller batches
- Manually tweak your graph, e.g. fusing tasks (hard, but the most promising)

Yesterday @jcrist and I had an interesting chat about an idea that might help out. The idea was to have an explicit "fuse all upstream tasks" function. Maybe that will help us out a bit.

At this point, several institutions have significant amounts of money would could potentially throw at this problem, which is a serious roadblock for scaling dask. I would welcome ideas from the dask devs about how that money could be effectively deployed to fix this issue at its root cause.

**Comment by takluyver** [comment]
Thanks! I have observed that throwing more resources at the problem sometimes makes it go away. I'll explore the other approaches you mention.

I also work at such an institution - [European XFEL](https://www.xfel.eu/) - and my immediate boss is a big fan of supporting the ecosystem of open source tools. So if there were a concrete plan for how money could be applied to make a substantial improvement in this area, I'd push for EuXFEL to contribute. (Obviously this is just me talking, and I can't promise anything)

**Comment by dougiesquire** [comment]
@takluyver, depending on your HPC system and workflows, another potential (but also not very satisfactory) workaround may be to run your jobs using a **single** (multi-core) worker. This has helped me get through problematic jobs in the past.

**Comment by mrocklin** [comment]
> At this point, several institutions have significant amounts of money we could potentially throw at this problem, which has emerged as a serious roadblock for scaling dask. I would welcome ideas from the dask devs about how that money could be effectively deployed to fix this issue at its root cause.

I apologize generally for not being able to resolve all of these problems recently.  As Dask grows it gets harder to track everything personally.  Money definitely helps to prioritize things though :)  Thank you @rabernat for bringing this up.  I'm starting to take on support contracts at Coiled, and hire some people to help work on them.  Money goes a long way to giving us more bandwidth, especially on problems like this one that require some dedicated thinking.  @rabernat @takluyver I'll reach out by e-mail.

**Comment by rabernat** [comment]
I thought I would mention that this sort of problem is currently being discussed on the Pangeo discourse forum: https://discourse.pangeo.io/t/best-practices-to-go-from-1000s-of-netcdf-files-to-analyses-on-a-hpc-cluster/588/9
Folks may find some relevant points in that thread.

**Comment by itamarst** [comment]
This seems like a good place to bring up a new option that I haven't seen discussed: Linux memory pressure statistics (https://www.kernel.org/doc/html/latest/accounting/psi.html). This is a fairly new metric, and it allows a fundamentally different approach to memory pressure.

Basically, it tells you "this is how much time you've spent waiting for the memory subsystem". So as memory resources get low, you start getting pressure feedback as swapping increases.

This means you don't have to try as hard to come up with estimates of memory usage in advance, you can be reactive to actual usage patterns as the code runs. You can imagine everything from "we should drop this cached data and recalculate" to load shedding to dynamically adjusting chunk sizes.

https://www.youtube.com/watch?v=cSJFLBJusVY is a good talk about how Facebook is using it, at a different level of abstraction than Distributed would, but the same approach (moving from overprovisioning to reacting to actual usage) would likely be quite useful in Distributed as well.

Other notes: there's also CPU and I/O (i.e. disk, I think?) backpressure info, and you can get the info per cgroup (==container).

**Comment by apatlpo** [comment]
Hi all, thank you very much for exposing this challenge @rabernat !
The issue was very helpful to our team.

I am not sure this is the right place for it but don't know where to push it otherwise.
I deeply apologize if this is not the place to report on this.

We have been using for some time an helper method in order to sequentially process large datasets along embarrassingly parallel dimensions and I'd like to share it: [here](https://gist.github.com/apatlpo/2194e1fe0187c431678db28eae8211d4)

The method has proven to be very valuable with time and the hope is that it can help other people.
There are most likely other approaches to do this and I would be very much looking forward to hear about it.

**Comment by gjoseph92** [comment]
So #4892 helps with this _a lot_. On my laptop (8-core intel MPB, 32GB memory), @rabernat's original example at the top of of this issue takes:
- `main`: **43 minutes**, spills ~130GB data to disk, makes my system's memory and makes it largely unresponsive
- #4892: **3.3 minutes**, no spilling, reasonable memory usage

main ([performance report](https://gistcdn.githack.com/gjoseph92/67e270c9fe891650a75762b25e525881/raw/9fdfb9b393c96183bc5f0f47de62b7164cdea0df/backpressure-main.html)):
![Screen Shot 2021-06-25 at 6 31 53 PM](https://user-images.githubusercontent.com/3309802/123496220-30a03e80-d5e4-11eb-9421-d63adc03271d.png)
#4892 ([performance report](https://gistcdn.githack.com/gjoseph92/67e270c9fe891650a75762b25e525881/raw/9fdfb9b393c96183bc5f0f47de62b7164cdea0df/backpressure-colocation.html)):
![Screen Shot 2021-06-25 at 6 32 06 PM](https://user-images.githubusercontent.com/3309802/123496223-3138d500-d5e4-11eb-8c98-3ae9aad44205.png)



I want to give a big shout-out to @JSKenyon, who so clearly identified the problem in #4864, and @fjetter who started us towards a solution. Dask had been scheduling the initial data-creation tasks in a very bad way, leading to huge amounts of unnecessary data transfer, and therefore data duplication, and therefore ballooning memory usage.

I made two changes before running these trials:
1. Added a `.copy()` at the end of the "pretend memory-reducing function":
    ```python
    def my_custom_function(f):
        # a pretend custom function that would do a bunch of stuff along
        # axis 0 and 2 and then reduce the data heavily
        return f.ravel()[::15][None, :].copy()  # <-- key: copy the slice to release `f`
    ```

    **This is essential**, because the small slice of `f` we were returning was just a view of `f`'s memory. And that meant all of `f`'s memory had to stick around—so the memory-reducing function was not reducing memory at all! The [NumPy docs](https://numpy.org/doc/stable/reference/arrays.indexing.html#basic-slicing-and-indexing) even mention this:

    > Care must be taken when extracting a small portion from a large array which becomes useless after the extraction, because the small portion extracted contains a reference to the large original array whose memory will not be released until all arrays derived from it are garbage-collected. In such cases an explicit `copy()` is recommended.

    This is unintuitive behavior, but I wonder how often it happens unknowingly with Dask, and if it plays a role in real-world cases. I wonder if we could add logic to Dask to warn you of this situation, where the memory backing an array result is, say, 2x larger than memory needed for the number of elements in that array, and suggest you make a copy.

1. Ran under jemalloc to avoid the [unreleased memory issue](https://distributed.dask.org/en/latest/worker.html#memory-not-released-back-to-the-os) (@crusaderky has explained this clearly in the Dask docs):

    ```bash
    DYLD_INSERT_LIBRARIES=$(brew --prefix jemalloc)/lib/libjemalloc.dylib python backpressure.py
    ```
    It probably would have still worked without this, but would have spilled more and been messier.

It's also interesting that memory backpressure wouldn't have helped with this exact example snippet. Because of the view-vs-copy issue, that code was essentially trying to compute all 93 GiB of the original data into memory. No matter your order through the graph, that end result's just not possible.

Not to say that we won't ever need memory backpressure (xref #4891). But I wonder how many other cases that _feel_ like Dask is producing more data than it can consume actually involve other sly bugs like the ones here. (Both the view-vs-copy and the poor ordering of root tasks were sly bugs, on behalf of a user and of dask.) And if there are tools/visualizations to help identify these issues that would serve us better than adding more complexity (memory backpressure) to try to work around them.

I'd love to hear if folks who have also run into these memory problems could retry their workloads with this update and report back how things change?

**Comment by itamarst** [comment]
> But I wonder how many other cases that feel like Dask is producing more data than it can consume actually involve other sly bugs like the ones here. ... And if there are tools/visualizations to help identify these issues that would serve us better than adding more complexity (memory backpressure) to try to work around them.

For issues that involve allocating too much memory or holding on to too much memory, you can figure out the sources of peak memory usage using the [Fil memory profiler](https://pythonspeed.com/fil). Though for now you'll need to use the threaded or single-threaded Dask schedulers, it doesn't support subprocesses yet.

**Comment by gjoseph92** [comment]
I've tried rerunning a couple examples from this thread under the new PR. Performance is greatly improved.

#### @rabernat's canonical anomaly-mean example from https://github.com/dask/distributed/issues/2602#issuecomment-498718651
20-worker cluster with 2-CPU, 10GiB workers

- `main`: 624.51s
- #4967: 155.39s (4x speedup)
- #4967 + [`MALLOC_TRIM_THRESHOLD_=0`](https://distributed.dask.org/en/latest/worker.html#automatically-trim-memory): 93.42s (6x speedup)

@mrocklin originally said about this example:
> I don't think that there is a way to compute that example in small memory without loading data twice. You have to load all of the data in order to compute `clim`. And then you need that data again to compute `anom_mean`.
> 
> The scheduling ask here is "be willing to delete and recompute data" while the scheduling ask before was "prefer tasks that release memory, and be willing to hold off a bit on tasks that increase net data".

This is still true. However in this example, `data` was only 200GiB, which is a reasonable size for a cluster (mine had exactly 200GiB total memory). So this still won't work on your laptop. But on a cluster, when we place the initial chunks of `data` well so they don't have to be shuffled around later, we can actually compute this quite efficiently.

#### @dougiesquire's climactic mean example from https://github.com/dask/distributed/issues/2602#issuecomment-535009454
20-worker cluster with 2-CPU, 20GiB workers

- `main` + [`MALLOC_TRIM_THRESHOLD_=0`](https://distributed.dask.org/en/latest/worker.html#automatically-trim-memory): gave up after 30min and 1.25TiB spilled to disk (!!)
- #4967 + [`MALLOC_TRIM_THRESHOLD_=0`](https://distributed.dask.org/en/latest/worker.html#automatically-trim-memory): 230s

<img width="1792" alt="Screen Shot 2021-06-29 at 3 04 01 PM" src="https://user-images.githubusercontent.com/3309802/123868905-8da23a00-d8ed-11eb-9bf5-9ed8b083e9ef.png">

FWIW, even with the colocation PR, this struggled when workers only had 8GiB of memory. Bumping to 20GiB smoothed things out (well, not for `main`). It would still be really cool if dask/distributed could tell you what amount of memory per worker was necessary to compute a given graph.

**Comment by dougiesquire** [comment]
Wow, this is really terrific @gjoseph92 ! Thanks heaps. Looking forward to giving it a spin.

**Comment by Thomas-Moore-Creative** [comment]
@gjoseph92 - that left-hand dashboard looks sadly familiar but the one on the right looks really exciting! Fantastic.

> It would still be really cool if dask/distributed could tell you what amount of memory per worker was necessary to compute a given graph.

+1
@mrocklin how much crowd-sourcing do you think we need to make that happen? ;-)

**Comment by RichardScottOZ** [comment]
So 200GiB problem required 400GiB cluster memory to run smoothly?  Seems like what happens generally from experience?

Be interesting to try that same test in the single large machine case - e.g. run the above ona  512GB machine what happens?

**Comment by mrocklin** [comment]
> @mrocklin how much crowd-sourcing do you think we need to make that happen? ;-)

Have you filled out the dask survey yet?  dask.org/survey

A lot of what we're working on now is motivated by responses there.  Although this work itself comes from generous sponsorship from @rabernat and crew.

**Comment by Thomas-Moore-Creative** [comment]
(hopefully) obviously it was a cheeky comment @mrocklin - but your response got me thinking about how other orgs ( beyond @rabernat et al ) might try to help resource these efforts.

Thanks for all the hard work from all contributors.  Going to hit that survey now.

**Comment by mrocklin** [comment]
Tell all your friends!

**Comment by sjperkins** [comment]
Really well done by all involved!

**Comment by JSKenyon** [comment]
This does look like a huge improvement! I look forward to trying it out on my problems.

**Comment by dcherian** [comment]
EXciting!

> This is essential, because the small slice of f we were returning was just a view of f's memory. And that meant all of f's memory had to stick around—so the memory-reducing function was not reducing memory at all! 
> This is unintuitive behavior, but I wonder how often it happens unknowingly with Dask, and if it plays a role in real-world cases. I wonder if we could add logic to Dask to warn you of this situation, where the memory backing an array result is, say, 2x larger than memory needed for the number of elements in that array, and suggest you make a copy.

@gjoseph92 This happens all the time! See https://github.com/dask/dask/issues/3595 (especially https://github.com/dask/dask/issues/3595#issuecomment-449544992) @mrocklin proposed a solution here: https://github.com/dask/dask/issues/3595 . I use the `map_blocks(np.copy)` trick frequently

**Comment by gjoseph92** [comment]
@dcherian thanks for the xref! https://github.com/dask/dask/issues/3595#issuecomment-451355007 is exactly what I was thinking. (It wouldn't help with user code in `map_blocks` like this case specifically, but since that's advanced, maybe that's okay.) I might like to take this on.

**Comment by mrocklin** [comment]
From a Coiled resourcing perspective, +1 on spending time on the getitem
fix.  That seems cheap to implement and medium-high value.

**Comment by mrocklin** [comment]
I'm closing this out.  If folks have other issues here though then I'm happy to reopen.

@rabernat , thank you for opening this originally.  My apologies that it took so long to come to a good solution.

**Comment by rabernat** [comment]
Which dask release do users need to be running to get this feature?

**Comment by mrocklin** [comment]
I think anything  2021.07 or greater should work.  We're also releasing
2021.07.2 today if you wanted to wait a day :)

**Comment by rabernat** [comment]
I am working on a blog post to advertise the exciting progress we have made on this problem. As part of this, I have created a [example notebook](https://gist.github.com/rabernat/39d8b6a396e076d168c24167b8871c4b) that can be run in Pangeo Binder (including with Dask Gateway) with two different versions of Dask which attempts to reproduce the issue and its resolution.

New Dask version (2021.07.1): <https://binder.pangeo.io/v2/gh/pangeo-gallery/default-binder/39f7202?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgist.github.com%252Frabernat%252F39d8b6a396e076d168c24167b8871c4b%26urlpath%3Dtree%252F39d8b6a396e076d168c24167b8871c4b%252Fanomaly_std.ipynb%26branch%3Dmaster>

Older Dask version (2020.12.1): <https://binder.pangeo.io/v2/gh/pangeo-gallery/default-binder/b8d1c53?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgist.github.com%252Frabernat%252F39d8b6a396e076d168c24167b8871c4b%26urlpath%3Dtree%252F39d8b6a396e076d168c24167b8871c4b%252Fanomaly_std.ipynb%26branch%3Dmaster>

The notebook uses the following cluster settings, which is the biggest cluster we can reasonably share publicly on Pangeo Binder:
```
nworkers = 30
worker_memory = 8
worker_cores = 1
```

It includes both the "canonical anomaly-mean example example" with synthetic data from https://github.com/dask/distributed/issues/2602#issuecomment-498718651 (referenced by @gjoseph92 in https://github.com/dask/distributed/issues/2602#issuecomment-870930134) as well as a similar real-world example that uses cloud data.

I have found that the critical performance issues are largely resolved even in Dask 2020.12.1 when I use `options.environment = {"MALLOC_TRIM_THRESHOLD_": "0"}`. For the synthetic data example, the older Dask version (pre #4967) is actually a little bit faster. For the real-world data example, the new version is significantly faster (2 min vs. 3 min), but nowhere close to the 6x increases reported above. But if I don't set `MALLOC_TRIM_THRESHOLD_`, both clusters crash.  This leads me to conclude that, **for these workloads, `MALLOC_TRIM_THRESHOLD_` is much more important than #4967**.

My questions for the group are:
1. Is this interpretation correct? Or am I missing something?
2. If it is correct that `MALLOC_TRIM_THRESHOLD_` makes the biggest differences for these climatology-anomaly-style workloads, is there an alternative workflow that would better highlight the improvements of #4967 for the blog post?
3. Is there a downside to setting `MALLOC_TRIM_THRESHOLD_=0` on all of our clusters?

**Comment by rabernat** [comment]
I'm trying to reproduce the "dougiesquire climatic mean example" with the following additional code

```python
import pandas as pd

size = (28, 237, 48, 21, 90, 144)
chunks = (1, 1, 48, 21, 90, 144)  # made the chunks smaller by a factor of 2
arr = dsa.random.random(size, chunks=chunks)

items = dict(
    ensemble = np.arange(size[0]),
    init_date = pd.date_range(start='1960', periods=size[1]),
    lat = np.arange(size[2]).astype(float),
    lead_time = np.arange(size[3]),
    level = np.arange(size[4]).astype(float),
    lon = np.arange(size[5]).astype(float),
)
dims, coords = zip(*list(items.items()))

array = xr.DataArray(arr, coords=coords, dims=dims)
dset = xr.Dataset({'data': array})

%time dset['data'].groupby("init_date.month").mean(dim="init_date").compute();
```

The newer Dask version (2021.07.1) does appear to run a bit faster, but both fail at the very end of the computation with a dead cluster and a `StreamClosedError`. @gjoseph92 - could you share the exact code snipped you used for this example in the results reported above?

<details>

```
---------------------------------------------------------------------------
StreamClosedError                         Traceback (most recent call last)
/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/comm/tcp.py in read(self, deserializers)
    192         try:
--> 193             frames_nbytes = await stream.read_bytes(fmt_size)
    194             (frames_nbytes,) = struct.unpack(fmt, frames_nbytes)

StreamClosedError: Stream is closed

The above exception was the direct cause of the following exception:

CommClosedError                           Traceback (most recent call last)
<timed eval> in <module>

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/core/dataarray.py in compute(self, **kwargs)
    953         """
    954         new = self.copy(deep=False)
--> 955         return new.load(**kwargs)
    956 
    957     def persist(self, **kwargs) -> "DataArray":

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/core/dataarray.py in load(self, **kwargs)
    927         dask.compute
    928         """
--> 929         ds = self._to_temp_dataset().load(**kwargs)
    930         new = self._from_temp_dataset(ds)
    931         self._variable = new._variable

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/core/dataset.py in load(self, **kwargs)
    863 
    864             # evaluate all the dask arrays simultaneously
--> 865             evaluated_data = da.compute(*lazy_data.values(), **kwargs)
    866 
    867             for k, data in zip(lazy_data, evaluated_data):

/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/base.py in compute(*args, **kwargs)
    566         postcomputes.append(x.__dask_postcompute__())
    567 
--> 568     results = schedule(dsk, keys, **kwargs)
    569     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    570 

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/client.py in get(self, dsk, keys, workers, allow_other_workers, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   2741                     should_rejoin = False
   2742             try:
-> 2743                 results = self.gather(packed, asynchronous=asynchronous, direct=direct)
   2744             finally:
   2745                 for f in futures.values():

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/client.py in gather(self, futures, errors, direct, asynchronous)
   2018             else:
   2019                 local_worker = None
-> 2020             return self.sync(
   2021                 self._gather,
   2022                 futures,

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/client.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    859             return future
    860         else:
--> 861             return sync(
    862                 self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    863             )

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    324     if error[0]:
    325         typ, exc, tb = error[0]
--> 326         raise exc.with_traceback(tb)
    327     else:
    328         return result[0]

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/utils.py in f()
    307             if callback_timeout is not None:
    308                 future = asyncio.wait_for(future, callback_timeout)
--> 309             result[0] = yield future
    310         except Exception:
    311             error[0] = sys.exc_info()

/srv/conda/envs/notebook/lib/python3.8/site-packages/tornado/gen.py in run(self)
    760 
    761                     try:
--> 762                         value = future.result()
    763                     except Exception:
    764                         exc_info = sys.exc_info()

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1912                 else:
   1913                     self._gather_future = future
-> 1914                 response = await future
   1915 
   1916             if response["status"] == "error":

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/client.py in _gather_remote(self, direct, local_worker)
   1963 
   1964             else:  # ask scheduler to gather data for us
-> 1965                 response = await retry_operation(self.scheduler.gather, keys=keys)
   1966 
   1967         return response

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/utils_comm.py in retry_operation(coro, operation, *args, **kwargs)
    383         dask.config.get("distributed.comm.retry.delay.max"), default="s"
    384     )
--> 385     return await retry(
    386         partial(coro, *args, **kwargs),
    387         count=retry_count,

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/utils_comm.py in retry(coro, count, delay_min, delay_max, jitter_fraction, retry_on_exceptions, operation)
    368                 delay *= 1 + random.random() * jitter_fraction
    369             await asyncio.sleep(delay)
--> 370     return await coro()
    371 
    372 

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/core.py in send_recv_from_rpc(**kwargs)
    864             name, comm.name = comm.name, "ConnectionPool." + key
    865             try:
--> 866                 result = await send_recv(comm=comm, op=key, **kwargs)
    867             finally:
    868                 self.pool.reuse(self.addr, comm)

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/core.py in send_recv(comm, reply, serializers, deserializers, **kwargs)
    649         await comm.write(msg, serializers=serializers, on_error="raise")
    650         if reply:
--> 651             response = await comm.read(deserializers=deserializers)
    652         else:
    653             response = None

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/comm/tcp.py in read(self, deserializers)
    201             self._closed = True
    202             if not sys.is_finalizing():
--> 203                 convert_stream_closed_error(self, e)
    204         except Exception:
    205             # Some OSError or a another "low-level" exception. We do not really know what

/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/comm/tcp.py in convert_stream_closed_error(obj, exc)
    123         raise CommClosedError(f"in {obj}: {exc.__class__.__name__}: {exc}") from exc
    124     else:
--> 125         raise CommClosedError(f"in {obj}: {exc}") from exc
    126 
    127 

CommClosedError: in <closed TLS>: Stream is closed
distributed.client - ERROR - Failed to reconnect to scheduler after 30.00 seconds, closing client
_GatheringFuture exception was never retrieved
future: <_GatheringFuture finished exception=CancelledError()>
asyncio.exceptions.CancelledError
```

</details>

**Comment by RichardScottOZ** [comment]
Thanks Ryan.  Have not tested this on anything seriously large, but hopefully will soon.

**Comment by dcherian** [comment]
> but I feel like I'm still missing something.

This thread is a confusing mix of many issues. The `slice` vs `copy` issue is real, but doesn't affect groupby problems since xarray indexes out groups with a list-of-ints (always a copy) [except for resampling which uses slices].

-----

> result = dset['data'].groupby("init_date.month")

This example (from @dougiesquire  I think) has chunksize=1 along `init_date` which is daily frequency. Xarray's groupby construct a nice graph for this case so it executes well (it'll pick out all chunks with january data and apply `dask.array.mean`; this is embarassingly parallel). If it doesn't execute well, that's up to dask+distributed to schedule it properly. 

-----
The earlier example:
```
cat = intake.Catalog('https://raw.githubusercontent.com/pangeo-data/pangeo-datastore/master/intake-catalogs/ocean.yaml')
ds = cat.GODAS.to_dask()
print(ds)
salt_clim = ds.salt.groupby('time.month').mean(dim='time')
```
`ds.salt` has [chunksize=4](https://catalog.pangeo.io/browse/master/ocean/GODAS/) along time with a monthly mean in each timestep so there are 4 months in a block
Array | Chunk
-- | --
11.31 GB | 96.08 MB
(471, 40, 417, 360) | (4, 40, 417, 360)
119 Tasks | 118 Chunks
float32 | numpy.ndarray

So `groupby("time.month")` splits every block into 4 and we end up with a quadratic-ish shuffling-type workload that tends to work poorly (though I haven't tried this on latest dask)
![image](https://user-images.githubusercontent.com/2448579/142449578-26c2a5fe-dd21-49e6-b1ab-78c3859032e1.png)

-----

The best way to reduce the GODAS dataset *should* be this strategy: https://flox.readthedocs.io/en/latest/implementation.html#method-cohorts i.e. we index out months 1-4 which always occur together and map-reduce that. Repeat for all other cohorts (months 5-8, 9-12) and then concatenate for the final result. 

One could test this with `flox.xarray.xarray_reduce(ds.salt, ds.time.dt.month, func="mean", method="cohorts")` and potentially throw in `engine="numba"` for extra fun =) Who wants to try it?!

**Comment by RichardScottOZ** [comment]
Thanks @gjoseph92
